stan/math/prim/meta/plain_type.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef STAN_MATH_PRIM_META_PLAIN_TYPE_HPP
+#define STAN_MATH_PRIM_META_PLAIN_TYPE_HPP
+
+#include <stan/math/prim/meta/plain_type.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * Determines plain (non expression) type associated with \c T. For non \c Eigen
+ * types it is the input type.
+ * @tparam T type to determine plain type of
+ */
+template <typename T, typename Enable = void>
+struct plain_type {
+  using type = std::decay_t<T>;
+};
+
+template <typename T>
+using plain_type_t = typename plain_type<T>::type;
+
+/**
+ * Determines return type of calling .eval() on eigen expression. This is the
+ * same as \c plain_type, except if input is already a plain type. In such case
+ * eval return type is a const reference to plain type.
+ * @tparam T type to determine eval return type of
+ */
+template <typename T>
+struct eval_return_type {
+  using T1 = plain_type_t<T>;
+  using type = std::conditional_t<std::is_same<std::decay_t<T>, T1>::value,
+                                  const T1&, T1>;
+};
```

Comment:
idt I'm totally understanding the intent of this function i.e. why specifically is `type` a `const T1&` when the decayed type is the same as the `plain_type`

Reply:
If `std::is_same<std::decay_t<T>` is `true` than T is a plain type - either `Eigen::Matrix` or `Eigen::Array`. These two types have overloaded `.eval()` that just returns a const ref to `this`. As `eval_return_type` is intended to be used to hold return value from calling eval it can be a const ref in these cases. Always using `plain_type` whithout ref would work, but would make a copy in these cases.

Reply:
> These two types have overloaded .eval() that just returns a const ref to this. As eval_return_type is intended to be used to hold return value from calling eval it can be a const ref in these cases. Always using plain_type whithout ref would work, but would make a copy in these cases.

Ah okay cool cool I didn't know that about Eigen with `.eval()`. Should we put that in the docs of plaintype somewhere? (less it's already there and I missed it)

Reply:
I think the docs satisfy this